activation/nipost_test.go

```diff
@@ -807,6 +807,393 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
+		})
+		require.NoError(t, err)
+
+		// successful post exec
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(poetProver, addedPoetProver), // add both poet provers
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("completely changed poet service BEFORE round started -> register new poet", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		addedPoetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr2)
+		addedPoetProver.EXPECT().Proof(gomock.Any(), "").Return(
+			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
+		)
+
+		mclock := defaultLayerClockMock(ctrl)
+
+		postClient := NewMockPostClient(ctrl)
+		nonce := types.VRFPostIndex(1)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(
+			&types.Post{
+				Indices: []byte{1, 2, 3},
+			}, &types.PostInfo{
+				Nonce: &nonce,
+			}, nil,
+		)
+
+		postService := NewMockpostService(ctrl)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+		// successfully registered to removed poet
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(1 * time.Second),
+		})
+		require.NoError(t, err)
+
+		// successful post exec
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(addedPoetProver), // add new poet
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("1 poet added AFTER round started -> too late to register to added poet, get proof from known",
+		func(t *testing.T) {
+			db := localsql.InMemory()
+			const publishEpoch = 5
+			challenge := types.NIPostChallenge{
+				PublishEpoch: publishEpoch,
+			}
+			challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+			ctrl := gomock.NewController(t)
+
+			currLayer := types.EpochID(publishEpoch - 1).FirstLayer()
+			genesis := time.Now().Add(-time.Duration(currLayer) * layerDuration)
+
+			require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+
+			poetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr)
+			poetProver.EXPECT().Proof(gomock.Any(), "1").Return(
+				&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
+			)
+
+			addedPoetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr2)
```

Comment:
Same hare: let's avoid the `AnyTimes()` on `Submit()` EXPECT used in the `defaultPoetServiceMock()`. Btw, probably it would be best to just remove `defaultPoetServiceMock()` altogether (different PR).
```suggestion
			addedPoetProver := NewMockPoetService(ctrl)
			addedPoetProver.EXPECT().Address().Return(poetProverAddr2).AnyTimes()
```